pkg/policy/visibility.go

```diff
@@ -0,0 +1,181 @@
+// Copyright 2019 Authors of Cilium
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package policy
+
+import (
+	"fmt"
+	"regexp"
+	"strconv"
+	"strings"
+
+	"github.com/cilium/cilium/pkg/u8proto"
+)
+
+var annotationRegex = regexp.MustCompile("^((<(Ingress|Egress)/([0-9]{1,6})/(TCP|UDP|ANY)/([A-Za-z]{3,32})>,)+)$")
+
+func validateL7ProtocolWithDirection(dir string, proto L7ParserType) error {
+	switch proto {
+	case ParserTypeHTTP:
+		return nil
+	case ParserTypeDNS:
+		if dir == "Egress" {
+			return nil
+		}
+	case ParserTypeKafka:
+		if dir == "Ingress" {
+			return nil
+		}
+	default:
+		return fmt.Errorf("unsupported parser type %s", proto)
+
+	}
+	return fmt.Errorf("%s not allowed with direction %s", proto, dir)
+}
+
+// NewVisibilityPolicy generates the VisibilityPolicy that is encoded in the
+// annotation parameter.
+// Returns an error:
+// * if the annotation does not correspond to the expected
+//   format for a visibility annotation.
+// * if there is a conflict between the state encoded in the annotation (e.g.,
+//   different L7 protocols for the same L4 port / protocol / traffic direction.
+func NewVisibilityPolicy(anno string) (*VisibilityPolicy, error) {
+	// Add a trailing comma so we can match the regex for now.
+	if anno[len(anno)-1] != ',' {
+		anno = anno + ","
+	}
+	if !annotationRegex.MatchString(anno) {
+		return nil, fmt.Errorf("annotation for proxy visibility did not match expected format %s", annotationRegex.String())
+	}
+
+	nvp := &VisibilityPolicy{
+		Ingress: make(DirectionalVisibilityPolicy),
+		Egress:  make(DirectionalVisibilityPolicy),
+	}
+
+	anSplit := strings.Split(anno, ",")
```

Comment:
Just a small note that the work being done on lines 69-116 could be achieved with regex groups.

Reply:
I've added a TODO here for now; would you prefer I take this up as part of this PR?

Reply:
Can be done later